jobserver/api.py

```diff
@@ -1,10 +1,56 @@
 from django_filters import rest_framework as filters
-from rest_framework import viewsets
+from rest_framework import serializers, viewsets
+from rest_framework.generics import ListAPIView
 
-from .models import Job, Workspace
+from .models import Job, JobRequest, Workspace
 from .serializers import JobShimSerializer, WorkspaceSerializer
 
 
+class WorkspaceSerializer2(serializers.ModelSerializer):
+    created_by = serializers.CharField(source="created_by.username", default=None)
+
+    class Meta:
+        fields = [
+            "name",
+            "repo",
+            "branch",
+            "db",
+            "created_by",
+            "created_at",
+        ]
+        model = Workspace
+
+
+class JobRequestAPIList(ListAPIView):
+    filterset_fields = ["backend"]
+    permission_classes = []
+    queryset = (
+        JobRequest.objects.filter(
+            jobs__completed_at__isnull=True,
+        )
```

Comment:
Just to double-check the logic on this as I can't remember off the top of my head what SQL Django generates here. We want two sorts of JobRequest: ones which have at least one Job where `completed_at` is null; and ones which have no associated Jobs at all. Will this match both conditions?

Reply:
Not even close 😂

I think I had the condition completely wrong in my head, will rewrite!

Reply:
After a lot of head scratching this actually works (!).  I cut it down to `JobRequest.objects.filter(jobs__completed_at__isnull=True).distinct()` which produces this SQL:

```
SELECT DISTINCT
                "jobserver_jobrequest"."id",
                "jobserver_jobrequest"."created_by_id",
                "jobserver_jobrequest"."workspace_id",
                "jobserver_jobrequest"."backend",
                "jobserver_jobrequest"."force_run_dependencies",
                "jobserver_jobrequest"."requested_actions",
                "jobserver_jobrequest"."callback_url",
                "jobserver_jobrequest"."sha",
                "jobserver_jobrequest"."identifier",
                "jobserver_jobrequest"."created_at"
           FROM "jobserver_jobrequest"
LEFT OUTER JOIN "jobserver_job" ON ("jobserver_jobrequest"."id" = "jobserver_job"."job_request_id")
          WHERE "jobserver_job"."completed_at" IS NULL
```

I've expanded the test (`test_jobrequestapilist_success` in `test_api.py`) to cover the four possible JobRequest states (all jobs completed, some jobs completed, no jobs completed, no jobs) and confirm we should list the latter 3 of those.

Reply:
Excellent!